### DISCUSSION TITLE
Component Re Render on Cache Reset with useMutation

### DISCUSSION THREAD

**Discussion by puvi034** [TanStack/query#1345] [Q&A]
Thanks for awesome library. I have a requirement where I directly use useMutation(mutateFn, options) and on success response , I set response in the QueryCache by creating new query key. Assume there are no instances of useQuery as my most of the calls are useMutation. In each mutation call I keep updating the queryCache. My question is does it re render the component where ever I have `useQueryCache` instance. I understand it re renders component has reference to `useQuery` instances.

**Comment by TkDodo** [comment]
> I understand it re renders component has reference to useQuery instances.

yes, that's true.`useQueryCache` is just react-context, it will not re-render if you update data inside the cache.

**Reply by puvi034** [reply]
I read somewhere in the document if active useQuery instance has reference to queryKey will render the components again. The question was more useQueryCache updates, does it re render component. If it does n't re render, we can't use this for state management solution.

**Reply by TkDodo** [reply]
let me re-phrase:

if you have `useQuery` that listens to `'my-key'`, it will re-render whenever you get new data for the query with that key.

it doesn't matter _how_ the data gets updated - could be because of `invalidateQueries` triggering a re-fetch, could be because you manually put some data in the cache due to `queryCache.setQueryData('my-key', myData)`. You have a subscription to that key and it will re-render.

> My question is does it re render the component where ever I have useQueryCache instance.

`useQueryCache` does not create a subscription, so this will not re-render:

```
function Component() {
  const queryCache = useQueryCache()
  
  <button onClick={() => queryCache.invalidateQueries('my-key')>Click</button>
}
```

hope I cleared up the confusion :)

**Reply by puvi034** [reply]
This was my understanding too. Sorry If my question isn't clear enough to you. I will try to re phrase my question.
In my app, there's only useMutation calls and I create cache manually right after I receive the response from the BE (onSuccess callback). I have multiple GQL mutation calls, which all returns the same object type and I keep updating cache so that other components having reference to useMuation can take advantage of updated cache.
The problem I do see here is useMutation call does not let me subscribe to the cache and I want my components to re render where-ever I have  useMutation instances (expect it work in the same way like useQuery). is it possible?

```
export const useContract = (Id: string) => {
  const queryCache = useQueryCache();
  const options = {
    onSuccess: (data: GqlResponse<CreateContractMutationResult>) => {
      // created this key for the first time and there was no useQuery before
      queryCache.setQueryData("contract", () => data);
    },
  };
  const [mutate, requestState] = useCreateContractMutation(options);
  useEffect(() => {
    mutate(cartId);
  }, []);

  return requestState;
};
```

**Reply by TkDodo** [reply]
It seems like it is possible with a manual subscription (https://react-query.tanstack.com/docs/api/#querycachesubscribe) but I am still not sure what you are trying to achieve. Maybe you have a runnable example (preferably codesandbox) that I can take a look at?

Put it that way: If you add data to the cache under the key `"contract"`, is there not a single instance where you are doing `useQuery('contract')` that creates a subscription to that data?

If this is really not the case and you are only using `useMutation`, then I wouldn't use react-query for that... Take the response from the mutation and put it into React-Context and subscribe via `useContext` - that would be much simpler. The advantage of react-query comes from subscribing to the data via `useQuery`, but the underlying assumption is that you have a data fetching function that will retrieve the data. Manually writing to the queryCache is just something you can do additionally to that, but it shouldn't be the primary means of filling the cache ...

**Reply by puvi034** [reply]
My case is well explained by you 2nd para. Thanks for your response.

**Reply by ZeeEssDoubleU** [reply]
@TkDodo, I have a further question.  There are instances where I just want `useQuery` to read cached data, without an associated data fetching function.  An instance like this would be just like what @puvi034 is talking about.  

For instance, here's a super contrived example.  Let's say I increment with `useMutation('counter', ...)`, then `onSuccess` I call `queryClient.setCachedData(['counterResult'])` with the response of `useMutation('counter, ...)`.  Now I just want to read the cached data with `useQuery('counterResult')`.

This seems to subscribe and return fresh data, but I keep getting an error that says `Missing queryFn`.  Is there any way to resolve this error without adding a fetch function?

**Reply by TkDodo** [reply]
> There are instances where I just want useQuery to read cached data, without an associated data fetching function.

but, why? you can just as easily use any other global state manager for that, and they are usually better suited. react-query is promise based, and you would always get one render cycle where your state is `loading` before you get your data. I'd put that data in zustand instead for example

**Reply by ZeeEssDoubleU** [reply]
My thought was that if data was already there, why move it elsewhere in the app?  That said, if moving it to the local state on success is best practice, that's what I'll do.  Thanks

**Comment by puvi034** [comment]
This question is also kind of related to my question.
https://github.com/tannerlinsley/react-query/discussions/1336
